### PR TITLE
add thiscall calling convention support

### LINKED ISSUE #42044: add thiscall calling convention
Trying to get servo/rust-bindgen working with win32 today, I ran into servo/rust-bindgen#541, where the `__thiscall` calling convention wasn't supported in bindgen.  (My use of bindgen doesn't actually generate calls to `__thiscall` functions, but the function representations need to support `__thiscall` AFAICT.)  Tracing back through the stack, it's not supported in `syntex_syntax`, and adding that support is AFAICT tricky until it's supported by rustc in some way.

Adding `__thiscall` was already proposed in #37 and/or #5853, but tangled up in much grander plans about FFI interfaces; this bug is more narrowly scoped to just adding support for the calling convention.  People can then experiment with it via bindgen and nightly Rust.  In this way, I think it's essentially like the existing support for `vectorcall`.

I think it's straightforward to do the work and am happy to sign up for doing it, but I'd like to know if there are strenuous objections to doing so.

### PR DESCRIPTION
**Pull request by froydnj**
This support is needed for bindgen to work well on 32-bit Windows, and also enables people to begin experimenting with C++ FFI support on that platform.

Fixes #42044.

### CODE REVIEW

**Comment by aidanhs** [comment]
Thanks for the PR @froydnj, we'll make sure it gets reviewed by @pnkfelix or another reviewer soon!

**Comment by carols10cents** [comment]
ping @pnkfelix, pinging you on irc too!

**Comment by arielb1** [comment]
r? @japaric 

Actually, I might go and review this myself tomorrow.

File: src/doc/unstable-book/src/language-features/abi-thiscall.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# `abi_thiscall`
+
+The tracking issue for this feature is: none.
```
Comment (nikomatsakis) [inline src/doc/unstable-book/src/language-features/abi-thiscall.md]: Can we create a tracking issue for this? I think it'd be fine to have one tracking issue covering all the various unstable ABIs. It'd also be nice to actually put a few words in there about what the purpose of this ABI is. If we're going to have a book we might as well write something in it. =)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
r=me with some text in the unstable book, bonus points for opening an issue, but I can do that later

**Comment by froydnj** [comment]
Thanks for the review!  Added a bit of text to the book, along with a link to the tracking issue.

r? @nikomatsakis

**Review by nikomatsakis — APPROVED** [review]

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/rust-lang/rust/pull/42220

**Pull request merged**